### PR TITLE
Multiple memory server implementation phase 2

### PR DESCRIPTION
**Pull request by gautham-bhat-k**
Added RPC interface for metadata manager

### CODE REVIEW

File: src/metadata/fam_metadata_manager_direct.cpp [hunk 1]
```diff
@@ -915,20 +928,21 @@ int FAM_Metadata_Manager::Impl_::metadata_insert_dataitem(
...
-
-        return ret;
-
     } else {
-
-        DEBUG_STDERR(dataitemId, "Region lookup failed");
-        return META_ERROR;
+        DEBUG_STDOUT(regionName, "region not found");
+        message << "Regioin does not exist";
+        throw Metadata_Service_Exception(METADATA_ERROR, message.str().c_str());
```
Comment (mashoodk) [inline src/metadata/fam_metadata_manager_direct.cpp:944]: You might want to consider using new exception macros. It has added advantage of printing line number etc.

File: src/cis/fam_cis_direct.cpp [hunk 2]
```diff
@@ -173,11 +167,7 @@ void Fam_CIS_Direct::destroy_region(uint64_t regionId, uint64_t memoryServerId,
     // cached KVS is freed in metadata_delete_region and calling
     // metadata_delete_region after DestroyHeap will result in SIGSEGV.
 
-    ret = metadataManager->metadata_delete_region(regionId);
-    if (ret != META_NO_ERROR) {
-        message << "Can not remove region from metadata service";
-        throw CIS_Exception(REGION_NOT_REMOVED, message.str().c_str());
-    }
+    metadataManager->metadata_delete_region(regionId);
```
Comment (mashoodk) [inline src/cis/fam_cis_direct.cpp:170]: Are you catching Metadata_Service_Exception (or its base class exception here or in callers of this routine (cis_server ? Or fam_allocator_client ? )

File: src/metadata/fam_metadata_manager_client.cpp [hunk 3]
```diff
@@ -35,7 +35,8 @@
 #include "common/fam_memserver_profile.h"
 using namespace openfam;
 
-#define STATUS_CHECK()                                                         \
+#if 0
```
Comment (mashoodk) [inline src/metadata/fam_metadata_manager_client.cpp:38]: #if 0 ??

File: test/setup.sh [hunk 4]
```diff
@@ -46,7 +46,7 @@ pkill metadataserver
 cd src
 echo "Starting Memory Server..."
 ./memoryserver -m ${1} -r ${2:-8787} -l ${3:-7500} -p ${4:-sockets} &
-#echo "Starting Metadata Server..."
-#./metadataserver -a ${1} -r ${5:-8989} &
+echo "Starting Metadata Server..."
+./metadataserver -a ${1} -r ${5:-8989} &
```
Comment (mashoodk) [inline test/setup.sh:50]: Are we starting metadata server by default ? Shouldnt this be direct path.

File: src/cis/fam_cis_direct.cpp (hunk 2 above)
Comment (mashoodk) [reply src/cis/fam_cis_direct.cpp:171]: I think you need to change cis_server to catch any exception. It catches CIS_Exception, So metadata_exception will not be caught.

File: src/memory-node/metadata_server_main.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
+/*
+ * memory_server_main.cpp
```
Comment (mashoodk) [inline src/memory-node/metadata_server_main.cpp:2]: Filename and copyright

File: src/memory-node/metadata_server_main.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    signal(SIGTERM, signal_handler);
+#endif
+
+    metadataService = NULL;
+    try {
+        metadataService = new Fam_Metadata_Manager_Server(rpcPort, name);
+        metadataService->run();
+    } catch (Metadata_Service_Exception &e) {
+        if (metadataService) {
+            delete metadataService;
```
Comment (VrashiPonnappa) [inline src/memory-node/metadata_server_main.cpp:106]:
```suggestion
            delete metadataService;
            metadataService = NULL;
```

File: src/cis/fam_cis_direct.cpp [hunk 7]
```diff
@@ -75,7 +75,7 @@ void cis_direct_profile_dump() {
 
 Fam_CIS_Direct::Fam_CIS_Direct() {
     allocator = new Memserver_Allocator();
-    metadataManager = FAM_Metadata_Manager::GetInstance();
+    metadataManager = new Fam_Metadata_Manager_Direct();
```
Comment (VrashiPonnappa) [inline src/cis/fam_cis_direct.cpp:78]: My assumption is that the metadata direct and RPC will we added in next submit

**Pull request merged**